### DISCUSSION TITLE
Import existing UUID and created date

### DISCUSSION THREAD

**Discussion by dbhobbs** [ory/kratos#2803] [Q&A] [answered]
Would y'all be open to allowing the admin-only credential import API to override the [ID](https://github.com/ory/kratos/blob/v0.10.1/identity/identity.go#L72) and the [CreatedAt](https://github.com/ory/kratos/blob/v0.10.1/identity/identity.go#L130) fields?

**Comment by vinckr** [comment, marked as answer]
Hello Daniel, 
can you elaborate a bit more on the use case? 
Is it migrating an existing system and keeping the IDs?

It think overriding the ID is not really an option without significant breaking changes, not sure about the CreatedAt fields. 

Is using [the metadata feature](https://www.ory.sh/docs/kratos/manage-identities/managing-users-identities-metadata) an option for your use case? It is intended exactly for this - migrating a legacy system and keeping some of the data associated in your new identity schema. 

> The Identity ID can not be changed and can not be chosen. This ensures future compatibility and optimization for distributed stores such as CockroachDB.

**Reply by dbhobbs** [reply]
I'm importing a large number of users that already have UUIDv4 identifiers. Our identifiers _must_ be unique and immutable and we rely on the database to ensure uniqueness. I considered using a metadata field, but we'll be doing a lot of lookups by _our_ identifier, and I wonder what kind of performance that will see, if it's even possible.

We've looked at using Google Cloud Identity Provider (Firebase Auth) as our identity system, which allows us to override the user id and created at fields. But as an identity/auth system, it's getting a bit stale and we were hoping to use Kratos instead as we're already using Hydra and Oathkeeper.

**Reply by vinckr** [reply]
Hey, 
that makes sense yea. You could cache the metadata response or session (/whoami) if doing an API call on every lookup is not an option. 
We recently introduced a session caching feature to the managed Ory Cloud service and you can offload the performance issue on us. With self-hosted Ory you would have to build such a cache yourself. 

What will happen to new users? Will you create an extra UUID for them or use the Kratos UUID?

**Reply by dbhobbs** [reply]
New users will use the Identity ID that Kratos provides. Our current plan is to stand up (internal use only) a forked version of Kratos v0.10.1 that allows us to specify the ID and CreatedAt date during the import process. That will ensure that we avoid any performance issues and brittle ID translation. Initial tests have been successful. Though we are looking at making some more optimizations for importing large numbers of identities.

You're welcome to mark this discussion as answered.

**Reply by vinckr** [reply]
Thanks Daniel, very interesting. Feel free to keep us posted how it goes!
